File: gh-2516/src/main/java/com/example/gh2516/WalletTransaction.java
package com.example.gh2516;

import org.springframework.data.neo4j.core.schema.Id;
import org.springframework.data.neo4j.core.schema.Node;
import org.springframework.data.neo4j.core.schema.Relationship;

import java.math.BigDecimal;
import java.util.List;

@Node("WalletTransaction")
public abstract class WalletTransaction {

	@Id
	public String id;
	public BigDecimal value;

	@Relationship(type = "HAS")
	public List<TransactionStatusUpdate> walletTransactionStatus;

	@Relationship(type = "CHANGES")
	public WalletBalance balance;

	public Long createdAt;
	public Long updatedAt;
}

File: gh-2516-kt/src/main/kotlin/com/example/gh2516kt/WalletTransaction.kt
package com.example.gh2516kt

import org.springframework.data.neo4j.core.schema.Id
import org.springframework.data.neo4j.core.schema.Node
import org.springframework.data.neo4j.core.schema.Relationship
import java.math.BigDecimal
import java.time.Instant

@Node("WalletTransaction")
abstract class WalletTransaction(
    @Id
    var id: String?,
    var value: BigDecimal?,

    @Relationship(type = "HAS")
    var walletTransactionStatus: List<TransactionStatusUpdate>?,

    @Relationship(type = "CHANGES")
    var balance: WalletBalance?,

    var createdAt: Long? = Instant.now().toEpochMilli(),
    var updatedAt: Long?,
) {
    val currentStatus get() = walletTransactionStatus?.lastOrNull()
}
